File: 0526-Beautiful Arrangement/0526-Beautiful Arrangement.cpp
class Solution {
public:
    int countArrangement(int N) {
        vector<int> used(N);
        int count = 0;
        dfs(used, 1, N, count);
        return count;
    }
    
private:
    void dfs(vector<int>& used, int start, int N, int& count) {
        if (start == N + 1) {
            ++count;
            return;
        }
        
        for (int i = 1; i <= N; ++i) {
            if (!used[i - 1] && (i % start == 0 || start % i == 0)) {
                used[i - 1] = 1;
                dfs(used, start + 1, N, count);
                used[i - 1] = 0;
            }
        }
    }
};

// Search from back and get better performance
class Solution {
public:
    int countArrangement(int N) {
        vector<int> used(N);
        int count = 0;
        dfs(used, N, N, count);
        return count;
    }
    
private:
    void dfs(vector<int>& used, int start, int N, int& count) {
        if (start == 0) {
            ++count;
            return;
        }
        
        for (int i = N; i >= 1; --i) {
            if (!used[i - 1] && (i % start == 0 || start % i == 0)) {
                used[i - 1] = 1;
                dfs(used, start - 1, N, count);
                used[i - 1] = 0;
            }
        }
    }
};

File: 0501-0600/0526-Beautiful Arrangement/0526-Beautiful Arrangement.py
class Solution:
    def countArrangement(self, N: int) -> int:
        used = [0] * (N + 1)
        count = 0
        def dfs(start):
            nonlocal count
            if start == 0:
                count += 1
                return
            
            for i in range(N, 0, -1):
                if (not used[i]) and (i % start == 0 or start % i == 0):
                    used[i] = 1
                    dfs(start - 1)
                    used[i] = 0
                    
        dfs(N)
        return count


class Solution2:
    def countArrangement(self, N: int) -> int:
        nums = [i for i in range(1 + N)]
        count = 0
        def dfs(start):
            nonlocal count
            if start == 0:
                count += 1
                return
            
            for i in range(start, 0, -1):
                nums[i], nums[start] = nums[start], nums[i]
                if nums[start] % start == 0 or start % nums[start] == 0:
                    dfs(start - 1)
                nums[i], nums[start] = nums[start], nums[i]
                    
        dfs(N)
        return count
